### PR TITLE
Resolve lock leak during recursive delete

### PR DESCRIPTION
**Pull request by ZhuTopher**
These changes are being made to address the following leak message:
```
2021-10-23 18:17:29,940 ERROR AlluxioResourceLeakDetector - LEAK: LockResource.close() was not called before resource is garbage-collected. See https://docs.alluxio.io/os/user/stable/en/operation/Troubleshooting.html#resource-leak-detection for more information about this message.
Recent access records: 
Created at:
	alluxio.resource.LockResource.<init>(LockResource.java:46)
	alluxio.resource.LockResource.<init>(LockResource.java:70)
	alluxio.resource.RWLockResource.<init>(RWLockResource.java:39)
	alluxio.resource.RefCountLockResource.<init>(RefCountLockResource.java:54)
	alluxio.collections.LockPool.get(LockPool.java:210)
	alluxio.master.file.meta.InodeLockManager.lockInode(InodeLockManager.java:169)
	alluxio.master.file.meta.SimpleInodeLockList.lockAndAddInode(SimpleInodeLockList.java:211)
	alluxio.master.file.meta.SimpleInodeLockList.lockInode(SimpleInodeLockList.java:95)
	alluxio.master.file.meta.CompositeInodeLockList.lockInode(CompositeInodeLockList.java:63)
	alluxio.master.file.meta.LockedInodePath.traverse(LockedInodePath.java:448)
	alluxio.master.file.meta.LockedInodePath.traverseOrClose(LockedInodePath.java:386)
	alluxio.master.file.meta.LockedInodePath.lockChild(LockedInodePath.java:358)
	alluxio.master.file.meta.LockedInodePath.lockChild(LockedInodePath.java:342)
	alluxio.master.file.meta.InodeTree.gatherDescendants(InodeTree.java:984)
	alluxio.master.file.meta.InodeTree.gatherDescendants(InodeTree.java:990)
	alluxio.master.file.meta.InodeTree.getDescendants(InodeTree.java:968)
	alluxio.master.file.DefaultFileSystemMaster.delete(DefaultFileSystemMaster.java:1823)
	alluxio.master.file.PrivilegedFileSystemMaster.delete(PrivilegedFileSystemMaster.java:356)
	alluxio.master.file.FileSystemMasterClientServiceHandler.lambda$remove$17(FileSystemMasterClientServiceHandler.java:359)
	alluxio.RpcUtils.callAndReturn(RpcUtils.java:121)
	alluxio.RpcUtils.call(RpcUtils.java:83)
	alluxio.RpcUtils.call(RpcUtils.java:58)
	alluxio.master.file.FileSystemMasterClientServiceHandler.remove(FileSystemMasterClientServiceHandler.java:357)
	alluxio.grpc.FileSystemMasterClientServiceGrpc$MethodHandlers.invoke(FileSystemMasterClientServiceGrpc.java:2578)
	io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:182)
	io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)
	io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
	io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
	alluxio.security.authentication.ClientIpAddressInjector$1.onHalfClose(ClientIpAddressInjector.java:57)
	io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)
	io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
	io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
	alluxio.security.authentication.AuthenticatedUserInjector$1.onHalfClose(AuthenticatedUserInjector.java:67)
	io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:331)
	io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:797)
	io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	alluxio.concurrent.jsr.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1378)
	alluxio.concurrent.jsr.ForkJoinTask.doExec(ForkJoinTask.java:609)
	alluxio.concurrent.jsr.ForkJoinPool.runWorker(ForkJoinPool.java:1356)
	alluxio.concurrent.jsr.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:131)
2021-10-23 18:17:29,941 ERROR AlluxioResourceLeakDetector - Leak detected when alluxio.leak.detector.exit.on.leak set to true. Shutting down the JVM
```

### CODE REVIEW

**Comment by ggezer** [comment]
`  private void gatherDescendants(LockedInodePath inodePath, List<LockedInodePath> descendants) {
    Inode inode = inodePath.getInodeOrNull();
    if (inode == null || inode.isFile()) {
      return;
    }
    for (Inode child : mInodeStore.getChildren(inode.asDirectory())) {
      LockedInodePath childPath;
      try {
        childPath = inodePath.lockChild(child, LockPattern.WRITE_EDGE);
      } catch (InvalidPathException e) {
        // Child does not exist.
        continue;
      }
      descendants.add(childPath);
      gatherDescendants(childPath, descendants);
    }
  }`

In above method, `descendants.add(childPath);` might also throw with `OutOfMemory`. Could you please fix this that well?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 1]
```diff
@@ -430,13 +430,8 @@ public LockedInodePath lockInodePath(AlluxioURI uri, LockPattern lockPattern)
...
     LockedInodePath inodePath =
-        new LockedInodePath(uri, mInodeStore, mInodeLockManager, getRoot(), lockPattern, tryLock);
-    try {
-      inodePath.traverse();
-    } catch (Throwable t) {
-      inodePath.close();
-      throw t;
-    }
+        new LockedInodePath(uri, mInodeStore, mInodeLockManager, getRoot(), lockPattern, tryLock)
+          .traverseOrClose();
```
Comment (ZhuTopher) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java:434]: This code is functionally identical to the prior code, but we use `traverseOrClose()` to DRY-ify it.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 2]
```diff
@@ -431,12 +431,7 @@ public LockedInodePath lockInodePath(AlluxioURI uri, LockPattern lockPattern, bo
       throws InvalidPathException {
     LockedInodePath inodePath =
         new LockedInodePath(uri, mInodeStore, mInodeLockManager, getRoot(), lockPattern, tryLock);
-    try {
-      inodePath.traverse();
-    } catch (Throwable t) {
-      inodePath.close();
-      throw t;
-    }
+    inodePath.traverseOrClose();
```
Comment (ZhuTopher) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java:434]: This code is functionally identical to the prior code, but we use `traverseOrClose()` to DRY-ify it.

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java [hunk 3]
```diff
@@ -198,7 +198,14 @@ private void addLock(RWLockResource lock, LockMode mode) {
     if (!endsInWriteLock() && mode == LockMode.WRITE) {
       mFirstWriteLockIndex = mLocks.size();
     }
-    mLocks.add(lock);
+    try {
+      mLocks.add(lock);
+    } catch (OutOfMemoryError e) {
```
Comment (ggezer) [inline core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java:203]: is `Error` better?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 4]
```diff
@@ -970,7 +965,14 @@ private void gatherDescendants(LockedInodePath inodePath, List<LockedInodePath>
         // Child does not exist.
         continue;
       }
-      descendants.add(childPath);
+      try {
+        descendants.add(childPath);
+      } catch (OutOfMemoryError e) {
```
Comment (ggezer) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java:970]: is `Error` better?

**Review by ggezer — APPROVED** [review]

**Comment by ggezer** [comment]
Error: 2.203 [ERROR] /usr/src/alluxio/core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java:[434,14] traverseOrClose() has private access in alluxio.master.file.meta.LockedInodePath

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 2 above)
Comment (ZhuTopher) [reply core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Nevermind, `traverseOrClose()` is `private` so this change is not trivial. Reverted.

**Comment by ggezer** [comment]
alluxio-bot, merge this please

**Comment by Xenorith** [comment]
alluxio-bot, cherry-pick this to branch-2.7 please

**Pull request merged**